tools/distpartitioning/data_shuffle.py

```diff
@@ -273,29 +263,25 @@ def exchange_node_features(rank, world_size, node_feature_tids, ntype_gnid_map,
                 local_idx_partid = local_idx[cond]
 
                 if (gnids_per_partid.shape[0] == 0):
-                    node_feats_per_rank.append({feat_key : torch.empty((0,), dtype=torch.float)})
-                    global_nid_per_rank.append({feat_key : torch.empty((0,), dtype=torch.int64)})
+                    node_feats_per_rank.append(torch.empty((0,1), dtype=torch.float))
+                    global_nid_per_rank.append(np.empty((0,1), dtype=np.int64))
                 else:
-                    node_feats_per_rank.append({feat_key : node_feats[local_idx_partid]})
-                    global_nid_per_rank.append({feat_key : gnids_per_partid})
+                    if (len(list(node_feats[local_idx_partid].size())) == 1):
+                        node_feats_per_rank.append(node_feats[local_idx_partid])
+                    else:
+                        node_feats_per_rank.append(node_feats[local_idx_partid])
+                    global_nid_per_rank.append(torch.from_numpy(gnids_per_partid).type(torch.int64))
 
             #features (and global nids) per rank to be sent out are ready
             #for transmission, perform alltoallv here.
-            output_feat_list = alltoall_cpu_object_lst(rank, world_size, node_feats_per_rank)
-            output_feat_list[rank] = node_feats_per_rank[rank]
+            output_feat_list = alltoallv_cpu_data(rank, world_size, node_feats_per_rank, node_feats.dtype)
 
-            output_nid_list = alltoall_cpu_object_lst(rank, world_size, global_nid_per_rank)
-            output_nid_list[rank] = global_nid_per_rank[rank]
+            #output_nid_list = alltoall_cpu_object_lst(rank, world_size, global_nid_per_rank)
+            output_nid_list = alltoallv_cpu_data(rank, world_size, global_nid_per_rank, torch.int64)
 
             #stitch node_features together to form one large feature tensor
-            own_node_features[feat_key] = []
-            own_global_nids[feat_key] = []
-            for idx, x in enumerate(output_feat_list):
-                own_node_features[feat_key].append(x[feat_key])
-                own_global_nids[feat_key].append(output_nid_list[idx][feat_key])
-            for k in own_node_features.keys(): 
-                own_node_features[k] = torch.cat(own_node_features[k])
-                own_global_nids[k] = np.concatenate(own_global_nids[k])
+            own_node_features[feat_key] = torch.cat(output_feat_list)
+            own_global_nids[feat_key] = torch.cat(output_nid_list).numpy()
```

Comment:
why one is pytorch tensor and the other is numpy array?

Reply:
Node features are read from the file as tensors, this is how load_tensors function works -- it always returns tensors. The global-nids is a numpy array so this is used as function argument to alltoall_cpu_data this numpy array is explicity converted to tensor and once the results are received they are again converted back to numpy arrays from tensors.

Reply:
my comment is really why you want to maintain node features as pytorch tensors and nid as numpy arrays. this causes unnecessary complexity for the code. since you are using numpy arrays almost everywhere, why don't you convert it to numpy array?

Reply:
is the format conversion zero copy?

Reply:
Added a comment, per our discussion, in the docstring.